File: backend/F1Fantasy/Controllers/ResultController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;
using F1Fantasy.Models;
using F1Fantasy.Services;

namespace F1Fantasy.Controllers;

[Authorize]
[ApiController]
[Route("api/[controller]")]
[EnableRateLimiting("read")]
public class ResultController : ControllerBase
{
    private readonly ResultService _resultService;
    private readonly ILogger<ResultController> _logger;

    public ResultController(ResultService resultService, ILogger<ResultController> logger)
    {
        _resultService = resultService;
        _logger = logger;
    }

    [HttpGet("season/{season}")]
    public async Task<ActionResult<IEnumerable<RaceWithResults>>> GetResultsBySeason(string season)
    {
        try
        {
            _logger.LogInformation("GET /api/result/season/{Season} - Fetching results for season", season);
            var results = await _resultService.GetResultsBySeasonAsync(season);
            _logger.LogInformation("Successfully retrieved results for {Count} races in season {Season}", 
                results.Count(), season);
            return Ok(results);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching results for season {Season}", season);
            throw;
        }
    }

    [HttpGet("season/{season}/round/{round}")]
    public async Task<ActionResult<RaceWithResults>> GetResultsByRace(string season, string round)
    {
        try
        {
            _logger.LogInformation("GET /api/result/season/{Season}/round/{Round} - Fetching results for race", 
                season, round);
            var result = await _resultService.GetResultsByRaceAsync(season, round);
            
            if (result == null)
            {
                _logger.LogWarning("No results found for season {Season}, round {Round}", season, round);
                return NotFound(new { message = $"No results found for season {season}, round {round}" });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} results for season {Season}, round {Round}", 
                result.Results?.Count ?? 0, season, round);
            return Ok(result);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching results for season {Season}, round {Round}", season, round);
            throw;
        }
    }

    [HttpGet("season/{season}/round/{round}/driver/{driverId}")]
    public async Task<ActionResult<Result>> GetResultByDriver(string season, string round, string driverId)
    {
        try
        {
            _logger.LogInformation("GET /api/result/season/{Season}/round/{Round}/driver/{DriverId} - Fetching result", 
                season, round, driverId);
            var result = await _resultService.GetResultByDriverAsync(season, round, driverId);
            
            if (result == null)
            {
                _logger.LogWarning("No result found for season {Season}, round {Round}, driver {DriverId}", 
                    season, round, driverId);
                return NotFound(new { message = $"No result found for driver {driverId} in season {season}, round {round}" });
            }
            
            _logger.LogInformation("Successfully retrieved result for driver {DriverId} in season {Season}, round {Round}", 
                driverId, season, round);
            return Ok(result);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching result for season {Season}, round {Round}, driver {DriverId}", 
                season, round, driverId);
            throw;
        }
    }

    [HttpGet("cached")]
    public async Task<ActionResult<IEnumerable<Result>>> GetCachedResults()
    {
        try
        {
            _logger.LogInformation("GET /api/result/cached - Fetching cached results");
            var results = await _resultService.GetCachedResultsAsync();
            
            if (!results.Any())
            {
                _logger.LogWarning("No cached results found. Database may be empty.");
                return Ok(new { 
                    message = "No cached results found. Try calling /api/result/season/2025 first to populate the cache.",
                    results = results 
                });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} cached results", results.Count());
            return Ok(results);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching cached results");
            throw;
        }
    }

    [HttpGet("season/{season}/sprint")]
    public async Task<ActionResult<IEnumerable<RaceWithResults>>> GetSprintResultsBySeason(string season)
    {
        try
        {
            _logger.LogInformation("GET /api/result/season/{Season}/sprint - Fetching sprint results for season", season);
            var results = await _resultService.GetSprintResultsBySeasonAsync(season);
            _logger.LogInformation("Successfully retrieved sprint results for {Count} races in season {Season}", 
                results.Count(), season);
            return Ok(results);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching sprint results for season {Season}", season);
            throw;
        }
    }

    [HttpGet("season/{season}/round/{round}/sprint")]
    public async Task<ActionResult<RaceWithResults>> GetSprintResultsByRace(string season, string round)
    {
        try
        {
            _logger.LogInformation("GET /api/result/season/{Season}/round/{Round}/sprint - Fetching sprint results for race", 
                season, round);
            var result = await _resultService.GetSprintResultsByRaceAsync(season, round);
            
            if (result == null)
            {
                _logger.LogWarning("No sprint results found for season {Season}, round {Round}", season, round);
                return NotFound(new { message = $"No sprint results found for season {season}, round {round}" });
            }
            
            _logger.LogInformation("Successfully retrieved {Count} sprint results for season {Season}, round {Round}", 
                result.SprintResults?.Count ?? 0, season, round);
            return Ok(result);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching sprint results for season {Season}, round {Round}", season, round);
            throw;
        }
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/controller/ResultController.java
package no.f1fantasy.controller;

import java.util.List;
import java.util.HashMap;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import no.f1fantasy.service.RaceWithResults;
import no.f1fantasy.service.ResultService;

@RestController
@RequestMapping("/api/result")
@PreAuthorize("isAuthenticated()")
@SuppressWarnings("null")
public class ResultController {

    private final ResultService resultService;

    public ResultController(ResultService resultService) {
        this.resultService = resultService;
    }

    @GetMapping("/season/{season}")
    public ResponseEntity<?> getResultsBySeason(@PathVariable String season) {
        try {
            List<RaceWithResults> results = resultService.getResultsBySeason(season);
            return ResponseEntity.ok(results);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching results for season: " + season));
        }
    }

    @GetMapping("/season/{season}/round/{round}")
    public ResponseEntity<?> getResultsByRace(@PathVariable String season, @PathVariable String round) {
        try {
            java.util.Optional<RaceWithResults> results = resultService.getResultsByRace(season, round);
            if (!results.isPresent()) {
                return ResponseEntity.status(404).body(errorMap("No results found for season " + season + ", round " + round));
            }
            return ResponseEntity.ok(results.get());
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching results for season " + season + ", round " + round));
        }
    }

    @GetMapping("/season/{season}/round/{round}/driver/{driverId}")
    public ResponseEntity<?> getResultByDriver(@PathVariable String season, @PathVariable String round, @PathVariable String driverId) {
        try {
            java.util.Optional<no.f1fantasy.entity.Result> result = resultService.getResultByDriver(season, round, driverId);
            if (!result.isPresent()) {
                return ResponseEntity.status(404).body(errorMap("No result found for driver " + driverId + " in season " + season + ", round " + round));
            }
            return ResponseEntity.ok(result.get());
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching result for driver " + driverId));
        }
    }

    @GetMapping("/cached")
    public ResponseEntity<?> getCachedResults() {
        try {
            List<no.f1fantasy.entity.Result> results = resultService.getCachedResults();
            if (results == null || results.isEmpty()) {
                Map<String, Object> response = new HashMap<>();
                response.put("message", "No cached results found");
                response.put("results", results);
                return ResponseEntity.ok(response);
            }
            return ResponseEntity.ok(results);
        } catch (Exception ex) {
            return ResponseEntity.status(500).body(errorMap("Error fetching cached results"));
        }
    }

    private Map<String, String> errorMap(String message) {
        Map<String, String> error = new HashMap<>();
        error.put("error", message);
        return error;
    }
}
